rewrite-java-11/src/main/java/org/openrewrite/java/Java11Parser.java

```diff
@@ -15,445 +15,188 @@
+            moduleClassLoader = new UnrestrictedModuleClassLoader(appClassLoader);
+
         }
-    }
 
-    public static class Builder extends JavaParser.Builder<Java11Parser, Builder> {
         @Override
         public Java11Parser build() {
-            return new Java11Parser(logCompilationWarningsAndErrors, classpath, classBytesClasspath, dependsOn, charset, styles, javaTypeCache);
+            lazyInitClassLoaders();
+
+            try {
+                //Load the parser implementation use the unrestricted module classloader.
+                Class<?> parserImplementation = Class.forName("org.openrewrite.java.isolated.ReloadableJava11Parser", true, moduleClassLoader);
+
+                Constructor<?> parserConstructor = parserImplementation
+                        .getDeclaredConstructor(Boolean.TYPE, Collection.class, Collection.class, Collection.class, Charset.class,
+                                Collection.class, JavaTypeCache.class);
+
+                parserConstructor.setAccessible(true);
+
+                JavaParser delegate = (JavaParser) parserConstructor
+                        .newInstance(logCompilationWarningsAndErrors, classpath, classBytesClasspath, dependsOn, charset, styles, javaTypeCache);
+
+                return new Java11Parser(delegate);
+            } catch (Exception e) {
+                throw new IllegalStateException("Unable to construct Java11Parser.", e);
+            }
         }
     }
 
-    private static class ByteArrayCapableJavacFileManager extends JavacFileManager {
-        private final List<PackageAwareJavaFileObject> classByteClasspath;
-
-        public ByteArrayCapableJavacFileManager(Context context,
-                                                boolean register,
-                                                Charset charset,
-                                                Collection<byte[]> classByteClasspath) {
-            super(context, register, charset);
-            this.classByteClasspath = classByteClasspath.stream()
-                    .map(PackageAwareJavaFileObject::new)
-                    .collect(toList());
+    /**
+     * Rewrite's JavaParser is reliant on java's compiler internal classes that are now encapsulated within Java's
+     * module system. Starting in Java 17, the JVM now enforces strong encapsulation of these internal classes and
+     * default behavior is to throw a security exception when attempting to use these internal classes. This classloader
+     * circumvents these security restrictions by isolating Rewrite's Java 17 parser implementation classes and then
+     * loading any of the internal classes directly from the .jmod files.
+     *
+     * NOTE: Any classes in the package "org.openrewrite.java.isolated" will be loaded into this isolated classloader.
+     */
+    private static class UnrestrictedModuleClassLoader extends ClassLoader {
+
+        static {
+            ClassLoader.registerAsParallelCapable();
```

Comment:
Note, I am registering this classloader as "parallel capable" and then using the same pattern used in ClassLoader for synchronizing the loadClass method. See https://github.com/openrewrite/rewrite/pull/1797/files#diff-8bc6211c47504aaff7724ac3f94894b07ccf54ac7e77a40fbe8e7e6a7bd48177R145

Without this, I was getting failures (only on the CI builds)